tests/sources/rest_api/integration/test_offline.py

```diff
@@ -100,6 +100,297 @@ def test_load_mock_api(mock_api_server):
+            id="one_empty_param",
+        ),
+        # Explicitly set page param gets ignored
+        pytest.param(
+            {
+                "path": "posts",
+                "params": {"sort": "desc", "locale": "en", "page": "100"},
+            },
+            {"sort": ["desc"], "locale": ["en"]},
+            id="explicit_page_param",
+        ),
+        # Incremental defined in endpoint
+        pytest.param(
+            {
+                "path": "posts",
+                "incremental": {
+                    "start_param": "since",
+                    "end_param": "until",
+                    "cursor_path": "id",
+                    "initial_value": 1,
+                    "end_value": 10,
+                },
+            },
+            {"since": ["1"], "until": ["10"]},
+            id="incremental_in_endpoint",
+        ),
+        # Incremental mixed with static params
+        pytest.param(
+            {
+                "path": "posts",
+                "params": {"sort": "desc", "locale": "en"},
+                "incremental": {
+                    "start_param": "since",
+                    "end_param": "until",
+                    "cursor_path": "id",
+                    "initial_value": 1,
+                    "end_value": 10,
+                },
+            },
+            {"sort": ["desc"], "locale": ["en"], "since": ["1"], "until": ["10"]},
+            id="incremental_in_endpoint_mixed_with_static",
+        ),
+        # Incremental defined in params
+        pytest.param(
+            {
+                "path": "posts",
+                "params": {
+                    "sort": "desc",
+                    "since": {
+                        "type": "incremental",
+                        "cursor_path": "id",
+                        "initial_value": 1,
+                    },
+                },
+            },
+            {"since": ["1"], "sort": ["desc"]},
+            id="incremental_in_params",
+        ),
+    ],
+)
+def test_single_resource_query_string_params(
+    mock_api_server, endpoint_params, expected_static_params
+):
+    page_counter = 1
+
+    def response_hook(response):
+        nonlocal page_counter
+        query_string_params = response.json().get("query_string_params")
+        expected_params = {**expected_static_params, "page": [str(page_counter)]}
+        assert query_string_params == expected_params
+        page_counter += 1
+
+    mock_source = rest_api_source(
+        {
+            "client": {
+                "base_url": "https://api.example.com",
+                "paginator": {
+                    "type": "page_number",
+                    "base_page": 1,
+                    "total_path": "total_pages",
+                },
+            },
+            "resources": [
+                {
+                    "name": "posts",
+                    "endpoint": {
+                        **endpoint_params,
+                        "response_actions": [response_hook],
```

Comment:
this is pretty cool. but `mock_api_server` is just a requests mock and you can get a call history from it: https://requests-mock.readthedocs.io/en/latest/history.html

Reply:
good point, thanks for the tip. fixed in 51d990392ddf28e6668d874df7f367247344e213